### PR TITLE
Add object type for StationId

### LINKED ISSUE #72: Add top-level type for station ID
(Continued from discussion in [#50](https://github.com/pflarue/ardop/pull/50#pullrequestreview-2159370820))

We should add a new data type to represent **Station ID**: callsign + SSID. Notional requirements:

* Canonicalize the string representation
  * `n0call` → `N0CALL`
  * `N0CALL-0` → `N0CALL`
  * `N0CALL-` → `N0CALL`

* Accept only callsigns 3 – 7 ASCII characters

* Accept only SSIDs in `A-Z` and numeric `1` to `15`

* Conversion to/from 6-byte wireline representation

* Comparison function (i.e., is this packet for me?)

The data type should probably contain:

1. The ASCII C string representation
2. The compressed 6-byte wireline representation

Everything except for (2) is strictly optional, but it is handy to have a type that can be `snprintf()`'d without further transforms.

Construction should probably involve a round-trip through the compressor to ensure it can be transmitted losslessly.

**NOTE**: for future protocol designs, I favor the [28-bit representation](https://bitbucket.org/widefido/js8call/src/24677a00343c2f3cc53f6762466a4b11893094f8/varicode.cpp#lines-942) from js8call, which seems to do a good job of representing all the world's callsigns. We cannot change the ARDOP wireline representation at this time, but I would encourage future protocols to adopt a more complete encoding.

### PR DESCRIPTION
**Pull request by cbs228**
Switch to a "smart object" representation for callsigns throughout all of ardopcf. This is like `Locator` (#93), but the impact of this PR is far greater since callsigns are much more frequent and critical than grid squares.

The new `StationId` object has the following advantages over the older processors:

* Full support for seven-character callsigns

* The **canonical representation** of callsigns is guaranteed. Station IDs are canonicalized via a round-trip through the compressor.
  
* Packages the following representations together:

  * Canonical full string (`N0CALL-A`)
  * Callsign string only (`N0CALL`)
  * SSID (`A`)
  * `Packed6` wireline representation

* Includes methods for printing and parsing arrays for `MYAUX` support.

* Fallible functions guarantee populated and valid StationId for all frames that require them.

* Catches errors that legacy code misses, including erroneously truncating `N0CALL-16` to `N0CALL-1`.

* Improved test coverage

The legacy (de)compressors are removed.

### CODE REVIEW

**Comment by cbs228** [comment]
7b53bc6839a263e808f1d884846ffce6c7391ad2: ardopc also includes a `CQ` callsign that doesn't appear anywhere in the spec.

File: src/common/Packed6.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+/**
+ * @brief Compress to eight characters from `str`
+ *
+ * Compresses up to eight ASCII characters in six bytes. If
+ * fewer than eight characters are provided, the remaining
+ * characters are padded with spaces (ASCII 32).
+ *
+ * Lowercase letters are silently converted to uppercase.
+ * Other invalid characters, or strings longer than 8 bytes,
+ * are replaced with spaces and will return an error.
```
Comment (pflarue) [inline src/common/Packed6.h]: Wording is confusing:  strings longer than 8 bytes are not replaced with spaced.

File: src/common/Packed6.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+#include <stddef.h>
+#include <stdint.h>
+
+#include "common/mustuse.h"
+
+/**
+ * @def PACKED6_MAX
+ * @brief Maximum uncompressed ASCII characters in a Packed6
+ */
+#define PACKED6_MAX		8
```
Comment (pflarue) [inline src/common/Packed6.h]: Using two tabs here seems like an odd choice unless it is part of some style guideline that I'm unaware of.  I would have used a single space, but am open to alternatives.

File: src/common/Packed6.c [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	intSum = 0;
+
+	for (size_t i = 4; i < 8; i++)
+	{
+		intSum = (64 * intSum) + work[i] - 32;
+	}
+
+	packed->b[3] = (uint8_t)(intSum >> 16) & 255;
+	packed->b[4] = (uint8_t)(intSum >> 8) & 255;
+	packed->b[5] = (uint8_t)intSum & 255;
```
Comment (pflarue) [inline src/common/Packed6.c]: Did you consider rolling these into an outer loop to reduce code duplication?

File: src/common/Packed6.h [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
+ * @brief Construct from packed bytes
+ *
+ * Creates a Packed6 from the given compressed `bytes`.
+ *
+ * @param[in]  bytes    Packed bytes. This method requires
+ *                      *exactly* six bytes from `bytes`.
+ * @param[in]  nbytes   Number of bytes to read from `bytes`.
+ *                      Must be exactly 6 (`sizeof(Packed6)`)
```
Comment (pflarue) [inline src/common/Packed6.h]: If nbytes must be exactly 6, then why include it as a parameter?  I see that in the implementation you check that this value that must be six is indeed six, but I'm not sure that this really adds any value.

Along with comments added at the implementation of this function, it seems like it might be reasonable to greatly simplify this function to the point that no return value is needed since it always succeeds.

File: src/common/Packed6.c [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+bool packed6_from_bytes(const void* bytes, size_t nbytes, Packed6* packed)
+{
+	if (nbytes != sizeof(packed->b))
+	{
+		bool ok = false;
+		ok &= packed6_from_str("", packed);
+		return ok;
+	}
```
Comment (pflarue) [inline src/common/Packed6.c]: It seems that the only reason ok is defined here and then combined with the return value of packed6_from_str()  is to satisfy ARDOP_MUSTUSE.  This raises the question of whether such a requirement introduces unnecessary complexity.  Unnecessary complexity can make code harder to understand, and thus more likely to contain bugs.  

This also appears to the be the only use of packed6_from_str().  packed6_from_str_slice("", 0, packed) could have been used here instead and packed6_from_str() could be eliminated. 

Thoughts?

File: src/common/Packed6.c [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+	for (size_t i = 3; i < 6; i++)
+	{
+		intSum = (intSum << 8) + packed->b[i];
+	}
+
+	work[4] = (uint8_t)((intSum >> 18) & 63) + 32;
+	work[5] = (uint8_t)((intSum >> 12) & 63) + 32;
+	work[6] = (uint8_t)((intSum >> 6) & 63) + 32;
+	work[7] = (uint8_t)(intSum & 63) + 32;
```
Comment (pflarue) [inline src/common/Packed6.c]: Did you consider rolling these into an outer loop to reduce code duplication?

File: src/common/Packed6.c [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+		bool ok = false;
+		ok &= packed6_from_str("", packed);
+		return ok;
+	}
+
+	memcpy(&packed->b[0], bytes, sizeof(packed->b));
+	return true;
+}
+
+bool packed6_to_str(const Packed6* packed, char* out, size_t outsize) {
```
Comment (pflarue) [inline src/common/Packed6.c]: Did you consider beginning by comparing outsize to PACKED6_MAX at the start of this function and immediately returning false on failure.  Then, work[] wouldn't be needed because out could be populated directly.

File: src/common/Packed6.h [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+ *
+ * Compresses up to `len` ASCII characters, `len <= 8`,
+ * in six bytes. If fewer than eight characters are provided,
+ * the remaining characters are padded with spaces (ASCII 32).
+ *
+ * Lowercase letters are silently converted to uppercase.
+ * Other invalid characters, or strings longer than 8 bytes,
+ * are replaced with spaces and will return an error.
+ *
+ * @param[in]  str      C string
```
Comment (pflarue) [inline src/common/Packed6.h]: Since this does not need to be a null terminated string, perhaps replace `C string` with `C string (null termination optional)`?

File: src/common/StationId.h [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+ */
+void stationid_make_cq(StationId* station);
+
+/**
+ * @brief Create a station ID from a string
+ *
+ * Accepts a string of the form "`N0CALL-15`" and parses it into a
+ * station ID. If this method returns zero, all fields of the
+ * station ID are populated and are ready for either user display
+ * or packet transmission.
```
Comment (pflarue) [inline src/common/StationId.h]: Please don't use the word "packet".  Earlier versions of ardop included AX.25 style packet communication, which has been eliminated.

File: src/common/StationId.h [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+ARDOP_MUSTUSE station_id_err
+stationid_from_str(const char* str, StationId* station);
+
+/**
+ * @brief Create a station ID from a bounded string
+ *
+ * Accepts a string of the form "`N0CALL-15`" and parses it into a
+ * station ID. If this method returns zero, all fields of the
+ * station ID are populated and are ready for either user display
+ * or packet transmission.
```
Comment (pflarue) [inline src/common/StationId.h]: Please don't use the word "packet". Earlier versions of ardop included AX.25 style packet communication, which has been eliminated.

File: src/common/StationId.h [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+ARDOP_MUSTUSE station_id_err
+stationid_from_str_slice(const char* str, size_t len, StationId* station);
+
+/**
+ * @brief Create a station ID from wireline bytes
+ *
+ * Creates a Packed6 from a byte buffer. `bytes` must have at
+ * least `nbytes >= 6` bytes available for reading. If this method
+ * returns zero, all fields of the station ID are populated and
+ * are ready for either user display or packet transmission.
```
Comment (pflarue) [inline src/common/StationId.h]: Please don't use the word "packet". Earlier versions of ardop included AX.25 style packet communication, which has been eliminated.

File: src/common/StationId.h [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+#include <stdint.h>
+
+#include "common/mustuse.h"
+#include "common/Packed6.h"
+
+/**
+ * @brief Error codes returned by Station ID functions
+ */
+typedef enum {
```
Comment (pflarue) [inline src/common/StationId.h:39]: My initial thought is that these error codes are counterproductive.  All the calling function needs to know is whether a the callsign is usable or not.  If one of these various problems was encountered, the only additional action required is to write a suitable log message that may help the user correct the problem.  It seems more practical to Implement this logging action when the problem is encountered than in each location where these functions are called from.

File: src/common/StationId.h [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+ *
+ * The canonical text representation of a station ID is
+ *
+ * ```c
+ * printf("%s-%s", station_id.call, station_id.ssid);
+ * ```
+ *
+ * with the callsign and SSID separated by a dash.
+ */
+typedef struct {
```
Comment (pflarue) [inline src/common/StationId.h:82]: Other than in the compression/decompression process, I don't believe that a callsign and its SSID are ever used separately.  So, why not have only two fields in StationID: str and wire, where str is the cannonical text representation?

File: src/common/StationId.h [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+	 * @brief Station Callsign
+	 *
+	 * A human-readable string like `N0CALL` which contains only
+	 * the callsign. ARDOP callsigns may contain 3 – 7 uppercase
+	 * ASCII letters and numbers.
+	 */
+	char call[8];
+
+	/**
+	 * @brief Station Service Set ID (SSID)
```
Comment (pflarue) [inline src/common/StationId.h]: In this context SSID is a Secondary Station ID, not a Service Set Identifier (as used for Wi-Fi networks, etc)

File: src/common/StationId.h [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+	 *
+	 * A human-readable string like `A` or `15` which contains
```
Comment (pflarue) [inline src/common/StationId.h:96]: If the callsign excluding the SSID is 7 characters, then the SSID may be only a single character.

File: src/common/StationId.c [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+ * Validate the `call`sign field of the given `station` ID.
+ * Callsigns which are too short, too long, or contain invalid
+ * characters are rejected. No attempt is made to enforce or
+ * validate a particular format (i.e., U.S. FCC conventions).
+ *
+ * Returns STATIONID_OK if the callsign is valid or
+ * STATIONID_ERR_INVALID_SSID if not.
+ */
+ARDOP_MUSTUSE station_id_err
+stationid_validate_callsign(const StationId* station);
```
Comment (pflarue) [inline src/common/StationId.c:31]: Is this function intended to eliminate the need for the existing CheckValidCallsignSyntax() function?

File: src/common/StationId.c [hunk 17]
```diff
@@ -0,0 +1,311 @@
+#include "common/StationId.h"
+
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+// minimum number of characters in a callsign
+#define CALLSIGN_MIN 3
```
Comment (pflarue) [inline src/common/StationId.c]: If we change this to 2, then we can eliminate the special case handling of CQ.  If we aren't enforcing any particular format, then 2 characters is as good as 3.  I verified that ARDOP_WIN correctly decodes arbitrary 2-character callsigns.  

I also checked, and ARDOP_WIN correctly decodes callsigns containing non-alphanumeric characters (contrary to the existing CheckValidCallsignSyntax() conditions).  This allows the use of / as required in some cases, though the 7-character limit is still limiting.

**Review by pflarue — COMMENTED** [review]
I reviewed only part of your pull request.

I think that the idea of using custom structs here is a good idea, though I added some questions about some of your other design choices.  Pretending that we are using an object oriented language, as you alluded to in your description of this pull request, is something I fully support.  If I thought we could do it in a reasonable amount of time and without breaking things, I'd be in favor of attempting a conversion from c to c++ so that we could take advantage of its object oriented structure as well as some of its inherently safer standard features.

While I agree that the changes in this pull request, and similar changes that I assume you will want to make continuing forward, are likely to be useful, I'm a bit frustrated by them and I'm unsure what to do about it.  So, I'll explain here how I am feeling.  

Providing the level of detailed review that I feel is appropriate takes a lot of time.  I'd rather spend that time working on visible improvements, trying to fix things that we know are causing problems, or implementing testing that may identify broken things we haven't yet noticed.  Fixing things that aren't broken is not high on my list of priorities while there are outstanding issues that I think deserve attention.  Some of the changes I have made recently don't necessarily fit into these categories, but they were changes whose effort seemed justifiable to improve the large scale maintainability of the project.

Part of the problem is that due to my own limitations, mentally compiling your work so that I fully understand it and can intelligently comment on it is a slow process.

One possible solution to this is for you to take over as primary developer of ardopcf.  I'd then be happy to submit pull requests to your project for the sort of changes that I think are important.  An alternative is that we agree to discuss the type of changes you plan to make in words or pseudo-code before you implement them.  Then, I limit my reviews of your pull requests to compiling and running tests (automated and manual) on the various hardware/OS combinations that we are trying to support.  This is contrary to my instincts, but it would probably allow us to each contribute to this project in the ways that will lead to the best rate of improvement.  I'm willing to consider these or another approach that you might suggest because I believe that your programming skills exceed my own, and because you seem genuinely interested in the success of this project.

[edit]  Improving the user documentation is another thing I'd be happy to spend my time on.

**Comment by cbs228** [comment]
> If I thought we could do it in a reasonable amount of time and without breaking things, I'd be in favor of attempting a conversion from c to c++

In my opinion, C++ just wraps every utter failure of C in a shiny chromium coating. Don't get me wrong, there are a *lot* of nice-to-haves… but if I was going to tackle a new-language rewrite, I'd rewrite it in Rust 🦀. It is a much more pleasant and robust language.

One shortcoming of Rust is that the build tools are very heavyweight. Rust has also some amazing reusable modules, but those modules will also add tons of transitive dependencies very quickly.

> Providing the level of detailed review that I feel is appropriate takes a lot of time. I'd rather spend that time working on visible improvements,

Yeah, I do a *lot* of code review. It's a tough skill to master. It's frequently no fun.

> One possible solution to this is for you to take over as primary developer of ardopcf.

I was kind of hoping to avoid that. I maintain so many things already that keeping up with them all is a challenge. I'm not much of a researcher, either.

* I have some DSP knowledge but very little experience with HF-specific challenges like severe multipath and fading. Modem design is a bit beyond me.

* I also don't get a lot of RF air time. My station is QRP portable and usually not assembled. The QRM is about as bad as it gets here. I've got maybe 8 watts continuous. I can reach the low-rate PSK modes, sometimes, but a QAM uplink is well out of my reach. Unless I make a substantial investment in upgrades, most of the modes aren't all that useful to me.

I have been tempted many times to embark on my own rewrite of ARDOP, but I don't think I really have anything new to bring to this scene. Incrementally better, perhaps, but not new.

> This is contrary to my instincts

I would encourage you to mark major changes in maintainership with something that requires user-facing action: for example, to change a project URL or package name. That prevents the changeover from passing silently, which is a serious security risk.

It's probably going to take a me a few days to get to the revisions, but it shall be done.

File: src/common/StationId.h (hunk 13 above)
Comment (cbs228) [reply src/common/StationId.h:82]: [`sendCWID()`](/ARDOPC/Modulate.c#L1425)  uses a bare callsign with no SSID, so keeping it separate will be handy.

File: src/common/StationId.h (hunk 15 above)
Comment (cbs228) [reply src/common/StationId.h:96]: I don't think so. Uncompressed callsigns get 8 ASCII characters, with the last always reserved for the SSID. The numeric SSID `15` is mapped to ASCII 63 (`?`) to make it fit. At least, this is what legacy ARDOPC does.

Can you check how ARDOP Windows encodes a `-15` SSID and if it works with 7-character callsigns? Some hex strings would be useful for the unit tests.

File: src/common/StationId.c (hunk 16 above)
Comment (cbs228) [reply src/common/StationId.c:31]: Yes. Moving the validation to construction time means we don't have to do it later.

File: src/common/Packed6.c (hunk 6 above)
Comment (cbs228) [reply src/common/Packed6.c]: I honestly just copy-pasted this from the original implementation. I decided to take another look at this algorithm and revise it in `compress_four_to_three()` and `decompress_three_to_four()`. I think this more clearly shows what is happening with the algorithm. If you don't like it, I can put it back.

File: src/common/Packed6.c (hunk 7 above)
Comment (cbs228) [reply src/common/Packed6.c]: Sounds good, done.

File: src/common/Packed6.c (hunk 3 above)
Comment (cbs228) [reply src/common/Packed6.c]: I honestly just copy-pasted this from the original implementation. I decided to take another look at this algorithm and revise it in `compress_four_to_three()` and `decompress_three_to_four()`. I think this more clearly shows what is happening with the algorithm. If you don't like it, I can put it back.

File: src/common/Packed6.h (hunk 2 above)
Comment (cbs228) [reply src/common/Packed6.h]: Fixed.

File: src/common/Packed6.h (hunk 1 above)
Comment (cbs228) [reply src/common/Packed6.h]: Fixed.

File: src/common/Packed6.h (hunk 8 above)
Comment (cbs228) [reply src/common/Packed6.h]: Done. I typically use `NUL` to refer to ASCII character 0 to disambiguate it from `NULL`, the invalid pointer.

File: src/common/Packed6.h (hunk 4 above)
Comment (cbs228) [reply src/common/Packed6.h]: > If nbytes must be exactly 6, then why include it as a parameter?

In APIs, I always try to pair an array with its capacity (for writes) or length (for reads). I think it is helpful for the caller to be explicit:

```c
Packed6 val;

packed6_from_bytes(&mybuf[0], 6, &val);
/* or maybe */
packed6_from_bytes(&mybuf[0], sizeof(Packed6), &val);
```

so that anyone reading that line knows exactly how many bytes are being read. It turns out that devs don't seem to like reading dev docs, so I'd rather not make people hunt for that information.

We could make the return value optional, but I'd want to convert it to an `assert()` statement. Asserts typically only work in debug-mode builds.

I also do something similar for `stationid_from_bytes()`, so whatever solution we come up with will be ported there as well.

File: src/common/StationId.h (hunk 14 above)
Comment (cbs228) [reply src/common/StationId.h]: Fixed.

File: src/common/StationId.h (hunk 9 above)
Comment (cbs228) [reply src/common/StationId.h]: "Packet" is often used interchangeably with "frame," but I will use "frame" instead. Fixed.

File: src/common/StationId.c (hunk 17 above)
Comment (cbs228) [reply src/common/StationId.c]: Widened the minimum to two. We now permit any packable character except space (and whitespace in general). Special constructors and tests for "CQ" remain because they will likely need to be special-cased.

> the 7-character limit is still limiting

Yes, IMO future protocols should adopt JS8call's [28-bit representation](https://bitbucket.org/widefido/js8call/src/24677a00343c2f3cc53f6762466a4b11893094f8/varicode.cpp#lines-942).

File: src/common/StationId.h (hunk 12 above)
Comment (cbs228) [reply src/common/StationId.h:39]: > It seems more practical to Implement this logging action when the problem is encountered

I concur, but I was really looking for a plan of action on #71 before I add more logging statements. We can work #71 first.

> All the calling function needs to know is whether a the callsign is usable or not.

The error codes are primarily for the tests: they let me assert which part of the code is rejecting the callsign. This is important to get good test coverage.

The API indicates that these functions return a value of zero for success and non-zero for error. This is very common in C, especially for system calls. In the driver code, I will almost certainly just check that the return value is zero.

**Comment by pflarue** [comment]
While I'm getting back used to working with C, and I think I could do the same for C++, everything I've read about Rust suggests a steep learning curve that I'm not interested in climbing at this time.

A good feature of Ardop for QRP use is that while the QRP station may only be able to reliably transmit using the slower more robust frame types, if the other station is using more power, the QRP station may receive data transmitted via faster frame types.  Of course, with both QRP and QRM, it is difficult.

Because we are constrained by the Ardop specification and not breaking compatibility with other Ardop implementations, incremental improvements are all we can hope for.  However, I believe that the combination of many of those incremental improvements can noticably improve reliability, performance, and ease of use.

Any change in maintainer would mean that the new master repository would be under somebody else's GitHub acct.  If somebody who I thought had the skills and interest in taking over wanted to do so, I'd then add a note to this repository directing people to that one.  However, it doesn't look like that is happening.  So, I'll continue.

Take your time.  As I wrote, I've got plenty of other things to focus on.  Right now I'm enhancing some diagnostic features that are helping me to better understand Issue #78.

File: src/common/Packed6.c (hunk 5 above)
Comment (cbs228) [reply src/common/Packed6.c]: > It seems that the only reason ok is defined here is to satisfy ARDOP_MUSTUSE.

Correct. Some versions of gcc will let you "use" like

```c
(void)i_totally_used_this;
```

but some ancient versions are too dumb for that.

Since `packed6_from_str("", packed)` has the same effect as zeroizing the Packed6, I can neatly sidestep this whole problem with a memset. This is what I have done in this revision.


> This also appears to the be the only use of packed6_from_str()

I also use it in the tests. I find it handy to have unbounded versions for use with string constants, but otherwise I prefer bounded.


> This raises the question of whether such a requirement introduces unnecessary complexity.

This is a fair criticism. Trying to side-step a `warn_unused_result` does occasionally result in some unnecessary branching. We can reduce this with internal-only `_unchecked` functions or by other workarounds. I will prioritize avoiding clutter that may result from this definition.

**Review by cbs228 — COMMENTED** [review]
> everything I've read about Rust suggests a steep learning curve

Rust is super hard until you "get it." When you "get it," you discover that Rust code can't look like C, Python, or Java, and that certain design patterns are necessary to reduce mutability and to control lifetimes. The two hardest things IMO are:

* Lifetime specifiers, which are less necessary these days but still crop up
* Generics (templates), and getting generic trait bounds correct. But that's hard in any language.

But by far the biggest disadvantage is having to start over from scratch. Re-using C in in Rust requires a lot of very messy interfacing and dealing with lots of build environments.

Anyway, a4029c0c38888ddb469240bd0cfa13a933c0a656 gets the easy stuff out of the way. There are a few threads that need more discussion.

I'm getting closer to having a working `Locator` object to store grid squares, but I'll omit that until we get the existing objects the way you want them.

File: src/common/Packed6.c (hunk 3 above)
Comment (pflarue) [reply src/common/Packed6.c]: `compress_four_to_three()` and `decompress_three_to_four()` look good.  However, I believe that the accompanying comments are misleading.  I don't believe that this is the same as HART Packed-ASCII.  If you want to give it a name, I believe that it does match DEC SIXBIT.

File: src/common/Packed6.c [hunk 18]
```diff
@@ -0,0 +1,139 @@
+#include "common/Packed6.h"
+
+#include <string.h>
+#include <stdio.h>
+
+// Converts vector of exactly four characters to a vector of exactly
+// three bytes using HART Packed-ASCII compression. The compressed
```
Comment (pflarue) [inline src/common/Packed6.c]: I don't think that this matches HART Packed-ASCII.

File: src/common/Packed6.c [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+	const uint8_t LOWERCASE_Z = (uint8_t)'z';
+	const uint8_t LOWER6 = (uint8_t)(0x3F);
+	const uint32_t BYTE = (uint32_t)(0xFF);
+
+	bool ok = true;
+	uint32_t pack = 0;
+
+	for (size_t ipos = 0; ipos < 4; ++ipos) {
+		// packed alphabet starts at ASCII 32 (space)
+		//  * lower-numbered characters become spaces
```
Comment (pflarue) [inline src/common/Packed6.c]: Spaces are used in place of anything less than 0x20 (space) as well as 0x60 and anything greater than 0x7A (z).  So your comment is incomplete and perhaps misleading.

File: src/common/Packed6.c [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+		} else {
+			// out of range
+			b = 0;
+			ok = false;
+		}
+
+		// take only the lower 6 bits and store them
+		// in the next-highest 6 bits of the pack word
+		pack = pack << 6;
+		pack |= ((uint32_t)(b & LOWER6));
```
Comment (pflarue) [inline src/common/Packed6.c]: Confirm this, but after the changes already made to b, I think that the two high bits are always zero.  So, applying `& LOWER6` is unnecessary.

File: src/common/Packed6.c [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+	for (size_t opos = 0; opos < 3; ++opos) {
+		compressed[2 - opos] = pack & BYTE;
+		pack = pack >> 8;
+	}
+
+	return ok;
+}
+
+// Converts vector of exactly three bytes to a vector of exactly
+// four ASCII characters using HART Packed-ASCII decompression. This
```
Comment (pflarue) [inline src/common/Packed6.c]: I don't think that this matches HART Packed-ASCII

File: src/common/Packed6.c [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+bool packed6_from_str(const char* str, Packed6* packed) {
+	const char* s = str ? str : "";
+
+	return packed6_from_str_slice(s, strnlen(s, PACKED6_MAX+1), packed);
+}
+
+bool packed6_from_str_slice(const char* str, size_t len, Packed6* packed) {
+	const char* s = str ? str : "";
```
Comment (pflarue) [inline src/common/Packed6.c]: `s` is not used.

File: src/common/Packed6.h (hunk 4 above)
Comment (pflarue) [reply src/common/Packed6.h]: Do as wish.

File: src/common/StationId.c [hunk 23]
```diff
@@ -0,0 +1,295 @@
...
+	out = stationid_uncompress(&station->wire, station);
+	if (out) {
+		stationid_init(station);
+		return out;
+	}
+
+	return out;
+}
+
+station_id_err stationid_from_bytes(const void* bytes, size_t nbytes, StationId* station) {
```
Comment (pflarue) [inline src/common/StationId.c]: What do you propose to do with the error result?  As I think I wrote before, upon encountering errors, I intend to write a suitable message to the debug log, but then continue on as best as possible.  If this is the intent, then why not handle such within the function?  If you are thinking something else, perhaps refusing a connection from someone with an unacceptable callsign (since perhaps this indicates a frame was incorrectly demodulated?), then please add comments indicating your intentions.

Your response to a similar question that I wrote earlier suggests that perhaps these return codes are primarily intended for testing purposes, and will thus be ignored during normal operation.  If this is your intent.  That seems reasonable.

File: src/common/StationId.h (hunk 12 above)
Comment (pflarue) [reply src/common/StationId.h:39]: See comments added to Issue #71.

Using the return values only for testing purposes seems reasonable.

File: src/common/StationId.h (hunk 13 above)
Comment (pflarue) [reply src/common/StationId.h:82]: I had not noticed that usage in sendcwid().  OK

File: src/common/StationId.h (hunk 15 above)
Comment (pflarue) [reply src/common/StationId.h:96]: I thought I remembered testing against ARDOP_WIN and finding that it truncated the result to a string of length 9 such that the second digit of the SSID was lost if the main portion of the callsign was 7 characters.  I was mistaken.  

Running ARDOP_Win with its associated GUI (which displays decoded IDFrame strings) allows be to see how it decodes IDFrames I send with ardopcf.  It is not as convenient to generate arbitrary frames with ARDOP_WIN for transmission to ardopcf.

Applying minor changes to the develop branch of ardopcf as it currently exists, including disabling CheckValidCallsignSyntax() and increasing CAL_BUF_SIZE from 10 to 11, I produce the following results:
[callsign string:  encoded hex value -> string decoded by ARDOP_WIN]
'A1A-1':  85 18 40 00 00 11 -> A1A-1
'N0CALL':  B9 08 E1 B2 C0 10 -> N0CALL
'N0CALL-A':  B9 08 E1 B2 C0 21 -> N0CALL-A
'N0CALL-Z':  B9 08 E1 B2 C0 3A -> N0CALL-Z
'N0CALL-[':  B9 08 E1 B2 C0 3B -> N0CALL-[
'N0CALL-0':  B9 08 E1 B2 C0 10 -> N0CALL
'LONGCAL-1':  B2 FB A7 8E 1B 11 -> LONGCAL-1
'LONGCAL-15':  B2 FB A7 8E 1B 1F -> LONGCAL-15
'LONGCAL-@':  B2 FB A7 8E 1B 20 -> LONGCAL-@
'LONGCAL-A':  B2 FB A7 8E 1B 21 -> LONGCAL-A
'CAFECAFE':  8E 19 A5 8E 19 90 -> CAFECAF
'MUCHTOOLON-G':  B7 58 E8 D2 FB D0 -> MUCHTOO
'MUCHTOOL-G':  B7 58 E8 D2 FB D0 -> MUCHTOO
'MUCHTOO-G':  B7 58 E8 D2 FB E7 -> MUCHTOO-G
'N0CALL-16':  B9 08 E1 B2 C0 11 -> N0CALL-1

**Review by pflarue — COMMENTED** [review]
Additional comments, questions, discussion are welcome, but please do not submit additional changes until this PR is ready to merge.  I don't want to spend much more time on this.

File: src/common/StationId.c (hunk 17 above)
Comment (pflarue) [reply src/common/StationId.c]: While such discussion is beyond the scope of this PR, I'll add the following:  A highly optimized callsign representation may be appropriate to a mode like JS8CALL whose payload capacity is very limited.  However, adding an additional few bytes to an ID Frame for an ARQ bulk data transfer mode would not have any significant impact on performance.  Thus, I would favor allowing greater flexibility over trying to keep this frame as short as possible.

**Comment by cbs228** [comment]
> please do not submit additional changes until this PR is ready to merge

I have the minimal fixes for the issues you've reported, but I assume from your comment that you don't want to see them until this is fully integrated in its final form. Correct?

> What do you propose to do with the error result?

* **Outgoing frames**: all errors will be prevented at the source. If the host attempts to use `MYCALL` or `GRIDSQUARE` to set an invalid callsign or grid square, the TNC will reject the attempt with the appropriate `FAULT` code. Frames with invalid callsigns will not be transmitted.

* **Incoming frames**: a frame with an invalid callsign should be considered invalid, discarded with a logged warning, and not acted upon any further. ID frames with invalid callsigns won't be reported as `IDF` on the host interface. ARQ calls with an invalid source callsign will elicit no reply. An invalid grid square will *not* invalidate the frame but will also not be reported to the host.

Basically, my intent is to treat these errors like they would be handled in IPv4: bad data → drop frame. Given the very wide acceptable criteria for a callsign, I do not think we will regret the breakage that this deliberately introduces.

> Your response to a similar question that I wrote earlier suggests that perhaps these return codes are primarily intended for testing purposes, and will thus be ignored during normal operation.

Correct, the calling functions will simply check if the error code is non-zero. Future revisions will add logging statements at the source of the error. That way, the calling function won't need to `switch` on it.

I'm going to see what the best solution to #71 will be before I come back to this.

**Comment by pflarue** [comment]
Yes, please continue until you have integrated the changes.

Regarding bad callsigns, your plan sounds good.  For from incoming frames, you may (or may not) have noticed a comment I added a while back in Decode4FSKConReq().  While I was updating the RS correction handling I noticed that these frames do include a few RS correction bytes, but do not use a CRC that can be used to verify whether or not the corrected results are actually correct.  So, I added a comment wondering whether the encoded callsigns should be checked for validity as a way to discard some incorrectly decoded ConReq frames.  This change you are working on should do that, which is good.

**Comment by cbs228** [comment]
2d7a1731197b8717343b1a322a299354da2e343f should fully integrate the new `StationId` type and remove the old (de)compressors. I have done basic loopback testing of this on Linux. ARQ sessions, pings, and ID frames appear to be working. I have tested connections between both 2d7a173-to-2d7a173 and 2d7a173-to-develop.

This version of `StationId` is a little different than in the previous review, but the basic concept is the same. Significant changes include:

* A new `str` field that contains the full callsign-SSID pre-formatted as a C string. This is used in too many places for us not to have it.

* New array-processing functions for MYAUX.

* Logging of wireline bytes that fail decode

* Prefer copying Packed6 to buffer with `stationid_to_buffer()`, which checks first that the StationId is populated and valid.

The basic idea remains the same.

This PR is broken out into several individual commits. The individual commits gradually roll out the new `StationId` to all parts of ardopcf. If you encounter a regression bug anywhere, I would appreciate a `git-bisect` to identify which commit is responsible.
